### PR TITLE
[SPARK-26825][SS] Fix temp checkpoint creation in cluster mode when default filesystem is not local.

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

There are situations where temporary checkpoint directory created by Spark. One example when one uses console sink. Such cases in the actual implementation `StreamingQueryManager` creates directory with `Utils.createTempDir` which will be passed to the appropriate `StreamExecution`. `StreamExecution` then does the following:
* Creates the directory again
* Resolves the provided directory

The problem comes when resolving happens. The `StreamingQueryManager` provided path doesn't contain `file://` scheme and because of this from local filesystem it can switch to HDFS for example (such case HDFS is the default filesystem).

In this PR I've added the following changes:
* Creating the directory only in `StreamExecution`
* `file://` scheme added to the directory
* As it was not clear that the checkpoint directory was not created because of permission issues I've added an exception when checkpoint directory doesn't exist and creation is not successful

## How was this patch tested?

Existing unit tests + started a query in client/cluster mode.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala
Comment (gaborgsomogyi) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: I've intentionally not used `toUri` here because of [#23733](https://github.com/apache/spark/pull/23733).

**Comment by gaborgsomogyi** [comment]
cc @jose-torres @dongjoon-hyun

**Comment by HeartSaVioR** [comment]
Could you rebase with current master? The code change itself looks good.

**Comment by gaborgsomogyi** [comment]
@HeartSaVioR thanks for reviewing it, I've merged the master.

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
It would be ideal to have UT preventing regression, but not sure whether it is easy to craft. Could you have a quick look whether we can add it or not?

**Comment by gaborgsomogyi** [comment]
Before the PR I've checked the possibilities but couldn't come up something which can be added as UT. What I was trying is to set default FS to something which is not `file` (for instance `hdfs`) and then check the resolved checkpoint path not changes the schema. This approach fails inside the resolve process. To resolve this somehow the filesystem has to be mocked where I didn't find solution. If you have some idea feel free because testing manually is brittle for long term.

**Comment by HeartSaVioR** [comment]
OK, I'll play with the PR when I get time. If I can achieve before merging this PR, I'll share the code. IMHO it doesn't need to be a blocker for merging.

**Review by HeartSaVioR — APPROVED** [review]
LGTM.

**Comment by HeartSaVioR** [comment]
@gaborgsomogyi 
Just raised PR on your fork regarding UT. https://github.com/gaborgsomogyi/spark/pull/1
Please review and merge if it makes sense. I'm OK to just raise follow-up PR on top of this patch when you would like to let me handle further reviews of my patch. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala
Comment (arunmahadevan) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: I dont think this handles cleaning up the tmp directory on shutdown which is already handled in `Utils.createTempDir`. I assume here we want the checkpoint location to be on the local file system and not in HDFS, so why not just re-use the Utils.createTempDir and just prefix "file://" to the canonical path to fix the issues with resolution.

**Review by arunmahadevan — CHANGES_REQUESTED** [review]
Since we want the tmp dir to be local, could just re-use the `Utils.createTempDir` to create the tmp checkpoint directory instead of a separate logic.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: I tend to agree with @arunmahadevan for leveraging `Utils.createTempDir`, but `file://` should be added as prefix of `root` parameter to `Utils.createTempDir` to prevent choosing default schema of filesystem and finally shutdown hook to delete directory from wrong filesystem.

Comment (arunmahadevan) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: I am not sure adding "file://" prefix to root parameter would cause the canonical path to return the final path with the same prefix. The attempted resolution with HDFS Filesystem happens later so we may be ok to just prepend the "file://" to canonical path.

**Comment by gaborgsomogyi** [comment]
`Utils.createTempDir` is not required for the following reasons:
* The separate logic is already there and works in a generic way but `Utils.createTempDir` works for local only
* Not required to create directory 2 times
* Shutdown hook doesn't help if the process killed with `kill -9`
* `StreamingQueryManager` responsibility is to come up with a directory and send it to `StreamExecution`
* `StreamExecution` responsibility to create and delete these directories

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: > I am not sure adding "file://" prefix to root parameter would cause the canonical path to return the final path with the same prefix.

That actually sounds serious to me and even sounds one reason to avoid.

**Comment by gaborgsomogyi** [comment]
@HeartSaVioR not yet checked all the details of your test but high level rocks. Many thanks investing your time! I think it would be better put it into this PR in order to help committers in the evaluation. Hope I can help you back the same way...

**Comment by HeartSaVioR** [comment]
@gaborgsomogyi Hey no worries! Happy to help. :)

**Comment by arunmahadevan** [comment]
@gaborgsomogyi thanks for the clarification, the `StreamExecution` apparently handles the deletion. However theres a slight behavior change here where earlier `Utils.CreateTempDir` would handle automatic deletion on shutdown and with the proposed one theres a possibility of the tmp dir being left over on shutdown. However I don't think its a big deal so I am ok with the proposed approach as long as we have validated creating the local dir via HDFS apis (the proposed one) does not have other differences with `Utils.CreateTempDir`.

**Comment by gaborgsomogyi** [comment]
@HeartSaVioR I've taken over your suggested test and made some minor modifications. Feel free to comment on this as well.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/StreamingCheckpointSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/StreamingCheckpointSuite.scala]: Nice to have verifying with checkpointDir!

Btw, if we know the path of checkpointDir, we can record actual path of request in MkdirRecordingFileSystem (like before) and let them just created, and find whether recorded directories have checkpointDir. This would make test still pass when Spark will be changed to create multiple directories including checkpointDir. 

This change may disable the impact of `fs.exist(checkpointDir)`, but we can determine which filesystem creates that directory so IMHO it's not a big deal.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/StreamingCheckpointSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/StreamingCheckpointSuite.scala]: I guess having SPARK-26825 as prefix helps to the future readers why the change was necessary. Please ignore of test name is too long to add it.

**Comment by jose-torres** [comment]
Do queries actually break without this change, or is the checkpoint just located in a different place? I think I'm missing context, because it seems correct for a temporary checkpoint to use the default filesystem instead of `file://` - after all, it is the default.

**Comment by HeartSaVioR** [comment]
Hmm... I might be missing to take a deep look at origin issue and hence wrong. I agree with @jose-torres statement.

We may need to re-analyze origin issue [SPARK-26825](https://issues.apache.org/jira/browse/SPARK-26825) again  - especially why file/directory creation fails in cluster mode. 

If temporary directory is correctly given, as it is running on YARN container, I would suspect why it fails to make a change on temporary directory.

```
*Cluster mode:*
java.io.tmpdir=/yarn/nm/usercache/root/appcache/application_1549064555573_0029/container_1549064555573_0029_01_000001/tmp/
createTempDir(namePrefix = s"temporary") => /yarn/nm/usercache/root/appcache/application_1549064555573_0029/container_1549064555573_0029_01_000001/tmp/temporary-47c13b28-14bd-4d1b-8acc-3e445948415e
```

```
19/02/01 12:53:14 ERROR streaming.StreamMetadata: Error writing stream metadata StreamMetadata(68f9fb30-5853-49b4-b192-f1e0483e0d95) to hdfs://ns1/data/yarn/nm/usercache/root/appcache/application_1548823131831_0160/container_1548823131831_0160_02_000001/tmp/temporary-3789423a-6ded-4084-aab3-3b6301c34e07/metadata
org.apache.hadoop.security.AccessControlException: Permission denied: user=root, access=WRITE, inode="/":hdfs:supergroup:drwxr-xr-x
```

```
scala> val fs = checkpointPath.getFileSystem(spark.sessionState.newHadoopConf())
fs: org.apache.hadoop.fs.FileSystem = DFS[DFSClient[clientName=DFSClient_NONMAPREDUCE_632752661_1, ugi=root (auth:SIMPLE)]]
scala> checkpointPath.makeQualified(fs.getUri, fs.getWorkingDirectory).toUri.toString
res1: String = hdfs://ns1/yarn/nm/usercache/root/appcache/application_1549064555573_0029/container_1549064555573_0029_01_000001/tmp/temporary-47c13b28-14bd-4d1b-8acc-3e445948415e
```

Could you find any weird thing here?

**Comment by gaborgsomogyi** [comment]
> Do queries actually break without this change, or is the checkpoint just located in a different place?

Yes, it fails to start. Additionally having an implementation where
* One part of the checkpoint is on local FS and the other part is on default FS (which is HDFS in this case). Users are checking the checkpoint data if the query fails which is in 2 different locations now.
* Assuming that locally created temp directory is writable on HDFS
is wrong in general.

**Comment by HeartSaVioR** [comment]
I'm not 100% sure which is right, but if we have to force creating temp checkpoint directory in local filesystem, we may need to know available temp directory of local filesystem instead of HDFS path what YARN provides us, because I'm not sure below directory can be created in local filesystem by any account running Spark (this is picked up from tmp dir in cluster mode):

```
/yarn/nm/usercache/root/appcache/application_1549064555573_0029/container_1549064555573_0029_01_000001/tmp/
```

**Comment by gaborgsomogyi** [comment]
What do you think @jose-torres ?

**Comment by gaborgsomogyi** [comment]
@HeartSaVioR `usercache` is local all the time on each node manager host, so creating it on HDFS is just bad.

**Comment by HeartSaVioR** [comment]
My bad. Now I realized the temporary directory YARN provides for container is local filesystem, not HDFS. If my understanding is right, then `creating the path in HDFS as it is` sounds me as a bug and the patch fixes it.

LGTM on the patch.

**Comment by gaborgsomogyi** [comment]
cc @zsxwing since Jose not responding.

**Comment by jose-torres** [comment]
My concern is that this change could break other scenarios that currently work. For example, I don't think stateful operations will work properly with a local FS checkpoint; executors won't be able to see the same checkpoint data.

**Comment by HeartSaVioR** [comment]
We're only dealing with temporary checkpoint here, so while I can see what @jose-torres mentioned, it is unlikely to happen in happy case and I haven't seen cache miss from running query. That could still happen in bad case like if some executors are becoming bad.

Btw, may be better to document somewhere that the checkpoint should be accessible from executor if query is stateful? I guess that's only thing where non-driver reads and writes into checkpoint directory, but I might be missing something.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala
Comment (vanzin) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: I'm not sure why `Utils.createTempDir` doesn't work here? What is the problem with `mkdir` being called twice?

Also I'm not sure I understand why you're not using `.toUri`. Even ignoring the PR you reference, which is closed, won't the string here look exactly the same?

```
scala> val f = new java.io.File("/tmp")
f: java.io.File = /tmp

scala> new Path(new Path(f.toURI()), "foo").toString
res1: String = file:/tmp/foo

scala> new Path(new Path(f.toURI()), "foo").toUri.toString
res2: String = file:/tmp/foo
```

Also the "file:" scheme doesn't generally like "//". Which is why you should use `File.toURI` instead of writing your own...

Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: > I'm not sure why Utils.createTempDir doesn't work here? What is the problem with mkdir being called twice?

`Utils.createTempDir` would work but an optional fix not to do it twice. The main problem is filesystem changes all of a sudden when default fs is not local. Please see the added test.

> Also I'm not sure I understand why you're not using `.toUri`.

```
scala> import org.apache.hadoop.fs.Path
import org.apache.hadoop.fs.Path

scala> val f = new java.io.File("/tmp/chk %#chk")
f: java.io.File = /tmp/chk %#chk

scala> new Path(new Path(f.toURI()), "foo").toString
res0: String = file:/tmp/chk %#chk/foo

scala> new Path(new Path(f.toURI()), "foo").toUri.toString
res1: String = file:/tmp/chk%20%25%23chk/foo
```

I know it can happen rarely but since it's discovered and we know it's better to prepare the code.

**Comment by gaborgsomogyi** [comment]
> My concern is that this change could break other scenarios that currently work. For example, I don't think stateful operations will work properly with a local FS checkpoint; executors won't be able to see the same checkpoint data.

@jose-torres Thank you for your time and good point. Not considered this scenario and agree that it may happen when executor becoming bad. I've double checked the sinks where temporary checkpoints are created and none of them guarantee fault tolerance. This change still fulfills this.

On the other hand if you still has concerns we can change the approach to create the temp checkpoint on default FS but then I would document that the directory configured by `java.io.tmpdir` system variable should be writable on the default filesystem. Please share your opinion.

@HeartSaVioR I agree with you and I would write out a warning when checkpoint (not just temp) is created on local FS (a good example is what Jose mentioned).

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: > Also the "file:" scheme doesn't generally like "//".

Not sure what you mean here (agree that manual URI creation is not the most appropriate way but considering the mentioned URI resolution this seemed the most reasonable way).

```
scala> val cpTempDir = new Path("file://" + tempDir + "/temporary1")
cpTempDir: org.apache.hadoop.fs.Path = file:/var/folders/t_/w90m85fn2gjb1v0c24wrfrxm0000gp/T/temporary1
```

Comment (vanzin) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: There have been bugs in the past caused by the host part (the "//") being included in file URIs. It's generally better to not add it since I don't believe it's even part of the respective RFC (although I didn't bother to go look).

Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: I see your point and agree. Let's wait to see the approach and together with that I'll modify this as well.

**Comment by gaborgsomogyi** [comment]
@zsxwing what do you think?

**Comment by gaborgsomogyi** [comment]
ping @jose-torres

**Comment by gaborgsomogyi** [comment]
ping @jose-torres

**Comment by gaborgsomogyi** [comment]
Having another view on this I think this has to be solved in a different way so closing.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala
Comment (BdLearnerr) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: @gaborgsomogyi , me confused , how to overwrite this Utils.createTempDir in spark-streaming code which uses spark-sql.2.4.1 version ?

**Pull request closed without merging**
